support custom modifiers

the current approach of parsing templates, requires all modifiers are present and registered.

find solution for that.

workaround would be that caller lists all modifiers and tsmarty2c registers empty modifiers by those names.

```
$ find templates -name '*.tpl.html' -o -name '*.tpl.text' -o -name '*.tpl.js' -o -name '*.tpl.xml' | xargs /home/glen/scm/smarty/tsmarty2c/bin/tsmarty2c.php -o ts.pot
Process templates/lookup_field.tpl.html
Process templates/error_message.tpl.html
Process templates/offline.tpl.text
Process templates/include/userstyle.tpl.html
Process templates/include/update_msg.tpl.html
Process templates/include/issue_field.tpl.html
Process templates/include/dropzone.tpl.html
Process templates/include/userscript.tpl.html
Process templates/include/notified_list.tpl.html
Process templates/include/issue_fields/assignee.tpl.html
Process templates/include/new_form.tpl.html
Process templates/include/attached_emails.tpl.html
Process templates/include/issue_fields.tpl.html
Process templates/include/rank_icon.tpl.html
Process templates/mail_queue.tpl.html

                                                                               
  [SmartyCompilerException]                                                    
  Syntax error in template "file:/home/glen/scm/eventum/eventum/localization/  
  workdir/templates/mail_queue.tpl.html"  on line 27 "<td width="20%" nowrap>  
  {$data[i].maq_queued_date|timeago}</td>" unknown modifier "timeago"          
                                                                               

extract [-o|--output OUTPUT] [--] <arguments> (<arguments>)...

make: *** [pot] Error 123
```

--- a/src/PluginLoader.php
+++ b/src/PluginLoader.php
@@ -1,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the smarty-gettext/tsmarty2c package.
+ *
+ * @copyright (c) Elan Ruusamäe
+ * @license BSD
+ * @see https://github.com/smarty-gettext/tsmarty2c
+ *
+ * For the full copyright and license information,
+ * please see the LICENSE and AUTHORS files
+ * that were distributed with this source code.
+ */
+
+namespace SmartyGettext;
+
+use Smarty;
+
+/**
+ * Provide dummy plugins for Smarty modifiers and blocks
+ *
+ * @see https://www.smarty.net/docs/en/api.register.default.plugin.handler.tpl
+ */
+class PluginLoader {
+	/**
+	 * List of plugin types to wrap with dummy handler.
+	 *
+	 * @var array
+	 */
+	private $providers = array(
+		Smarty::PLUGIN_MODIFIER => true,
+		Smarty::PLUGIN_BLOCK => true,
+		Smarty::PLUGIN_FUNCTION => true,
+	);
+
+	public function __invoke($name, $type, $template, &$callback) {
+		if (!isset($this->providers[$type])) {
+			return false;
+		}
+
+		// need to use static invocation
+		// because Smarty can't deal with dynamic calls.
+		$callback = array(__CLASS__, 'noop');
+
+		return true;
+	}
+
+	public static function noop() {
+	}
+}

--- a/src/PotFile.php
+++ b/src/PotFile.php
@@ -34,6 +34,7 @@ class PotFile {
 
 	public function __construct($outputFile) {
 		$smarty = new Smarty();
+		$smarty->registerDefaultPluginHandler(new PluginLoader());
 
 		$this->pofile = $this->getPoFile($outputFile);
 		$this->loader = new TokenLoader($smarty, $this->pofile);

--- a/src/PotFile.php
+++ b/src/PotFile.php
@@ -32,11 +32,17 @@ class PotFile {
 	/** @var PoFile */
 	private $pofile;
 
-	public function __construct($outputFile) {
+	/**
+	 * PotFile constructor.
+	 *
+	 * @param string $templateFile
+	 * @throws FileNotReadableException
+	 */
+	public function __construct($templateFile = null) {
 		$smarty = new Smarty();
 		$smarty->registerDefaultPluginHandler(new PluginLoader());
 
-		$this->pofile = $this->getPoFile($outputFile);
+		$this->pofile = $this->getPoFile($templateFile);
 		$this->loader = new TokenLoader($smarty, $this->pofile);
 		$this->parser = new TokenParser($smarty);
 	}

--- a/tests/ParserTest.php
+++ b/tests/ParserTest.php
@@ -1,0 +1,34 @@
+<?php
+
+/*
+ * This file is part of the smarty-gettext/tsmarty2c package.
+ *
+ * @copyright (c) Elan Ruusamäe
+ * @license BSD
+ * @see https://github.com/smarty-gettext/tsmarty2c
+ *
+ * For the full copyright and license information,
+ * please see the LICENSE and AUTHORS files
+ * that were distributed with this source code.
+ */
+
+namespace SmartyGettext\Test;
+
+use SmartyGettext\PotFile;
+use Symfony\Component\Finder\SplFileInfo;
+
+class ParserTest extends TestCase {
+
+	public function testParseUnknownModifier() {
+		$res = $this->parseTemplate(__DIR__ . '/data/modifier.tpl');
+		$this->assertNotNull($res);
+	}
+
+	private function parseTemplate($filename) {
+		$file = new SplFileInfo($filename, $filename, $filename);
+		$p = new PotFile();
+		$p->loadTemplate($file);
+
+		return $p->getOutput();
+	}
+}

--- a/tests/ParserTest.php
+++ b/tests/ParserTest.php
@@ -24,6 +24,11 @@ class ParserTest extends TestCase {
 		$this->assertNotNull($res);
 	}
 
+	public function testParseUnknownBlock() {
+		$res = $this->parseTemplate(__DIR__ . '/data/custom_block_single_tag.tpl');
+		$this->assertNotNull($res);
+	}
+
 	private function parseTemplate($filename) {
 		$file = new SplFileInfo($filename, $filename, $filename);
 		$p = new PotFile();

--- a/tests/data/custom_block_single_tag.tpl
+++ b/tests/data/custom_block_single_tag.tpl
@@ -1,0 +1,3 @@
+a template with custom block function
+
+hello {get_display_style element_name="quick_filter"} raisk

--- a/tests/data/modifier.tpl
+++ b/tests/data/modifier.tpl
@@ -1,0 +1,6 @@
+template containing modifier
+
+should not give parse error.
+https://github.com/smarty-gettext/tsmarty2c/issues/5
+
+<td width="20%" nowrap>{$data[i].maq_queued_date|timeago}</td>

